hawkbit-repository/hawkbit-repository-api/src/main/java/org/eclipse/hawkbit/repository/TargetQueryExecutionManagement.java

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) 2019 Bosch Software Innovations GmbH and others.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.hawkbit.repository;
+
+import org.eclipse.hawkbit.im.authentication.SpPermission;
+import org.eclipse.hawkbit.im.authentication.SpPermission.SpringEvalExpressions;
+import org.eclipse.hawkbit.repository.model.Target;
+import org.eclipse.hawkbit.repository.model.TargetFilterQuery;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.security.access.prepost.PreAuthorize;
+
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+import java.util.Collection;
+
+/**
+ * This interface is the central point which is responsible to execute queries
+ * and return update-targets for Hawkbit.
+ *
+ * By design it should be as minimal as possible in order to make it easy to
+ * plug-in different sources of {@linkplain Target Targets}.
+ * 
+ * For any use-case specific query use the query-/count-methods which limit the
+ * result to a predefined set of IDs.
+ */
+public interface TargetQueryExecutionManagement {
```

Comment:
Could you just extend the `TargetManagement` interface with the necessary methods, using the default method stub implementations instead of defining this class? In that case you won't need to adapt all of the classes that are using the target management already.

Reply:
The goal of this PR is to _seperate_ the query from the rest of the stuff done in TargetManagement in order to have it more easy to replace the source of the targets by an external device-management or ditto.
All methods in this new interface have been (with slightly different names) in the `TargetManagement` before moving them back would mean cancel the PR.

Reply:
I like the idea of @bogdan-bondar. `TargetManagement` has still methods related to queries/search e.g. `findByFilters(..)` or `findByQueryAndTag(...)`.
So I assume that (for now) `TargetQueryExecutionManagement` contains only the methods you need for your use-case i.e device management. 

Aren't you more flexible if you'll be able to change every aspect of `TargetManagement` at any time without modifying the consumer code? 

```java

class JpaTargetManagement implements TargetManagement {

    public Target findBy(...){
        // JPA call here
    }

    public void save(Target t){
        // JPA call here
    }
}

class MyExternalDeviceManagement implements TargetManagement {

    private JpaTargetManagement delegate; // if needed

    MyExternalDeviceManagement(JpaTargetManagement delegate){
        this.delegate = delegate;
    }

    public Target findBy(...){
        // call to external DM here
    }

    public void save(Target t){
        delegate.save(target); 
        // or save to external DM or both (what ever is required)
    }
}
```

Reply:
The idea was to seperate rollouts specific logic using entities which only exist in the "rollouts-world" like distribution-sets, rollouts, rollouts-groups and puer query operations.

Therefor finder-methods which join targets with somer rollouts-specific entity (like `findTargetsNotInRolloutGroup` stay in `TargetManagement`. 

We had already used Bogdans suggestion for an intermediate hack to get things going - but this was copy & paste the complete `JpaTargetManagement` and just replacing some methods. 

I would prefere to have the logic only in one place.